core/common/src/main/java/alluxio/exception/status/AlluxioStatusException.java

```diff
@@ -0,0 +1,164 @@
+ */
+
+package alluxio.exception.status;
+
+import alluxio.exception.AccessControlException;
+import alluxio.exception.AlluxioException;
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.BlockInfoException;
+import alluxio.exception.ConnectionFailedException;
+import alluxio.exception.DependencyDoesNotExistException;
+import alluxio.exception.DirectoryNotEmptyException;
+import alluxio.exception.FailedToCheckpointException;
+import alluxio.exception.FileAlreadyCompletedException;
+import alluxio.exception.FileAlreadyExistsException;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.exception.InvalidFileSizeException;
+import alluxio.exception.InvalidPathException;
+import alluxio.exception.InvalidWorkerStateException;
+import alluxio.exception.LineageDeletionException;
+import alluxio.exception.LineageDoesNotExistException;
+import alluxio.exception.NoWorkerException;
+import alluxio.exception.UfsBlockAccessTokenUnavailableException;
+import alluxio.exception.WorkerOutOfSpaceException;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.nio.file.attribute.UserPrincipalNotFoundException;
+import java.util.ConcurrentModificationException;
+import java.util.NoSuchElementException;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * An exception thrown by Alluxio. {@link #getStatus()} can be used to determine the represented
+ * class of error.
+ */
+public class AlluxioStatusException extends RuntimeException {
+  private static final long serialVersionUID = -7422144873058169662L;
+
+  private final ExceptionStatus mStatus;
+
+  /**
+   * @param status the status code for this exception
+   * @param message the exception message
+   */
+  public AlluxioStatusException(ExceptionStatus status, String message) {
+    super(message);
+    mStatus = status;
+  }
+
+  /**
+   * @param status the status code for this exception
+   * @param cause the cause of the exception
+   */
+  public AlluxioStatusException(ExceptionStatus status, Throwable cause) {
+    this(status, cause.getMessage(), cause);
+  }
+
+  /**
+   * @param status the status code for this exception
+   * @param message the exception message
+   * @param cause the cause of the exception
+   */
+  public AlluxioStatusException(ExceptionStatus status, String message, Throwable cause) {
+    super(message, cause);
+    mStatus = status;
+  }
+
+  /**
+   * @return the status code for this exception
+   */
+  public ExceptionStatus getStatus() {
+    return mStatus;
+  }
+
+  /**
+   * Converts checked Alluxio exceptions to Alluxio status exceptions.
+   *
+   * @param ae the Alluxio exception to convert
+   * @return the converted {@link AlluxioStatusException}
+   */
+  public static AlluxioStatusException fromAlluxioException(AlluxioException ae) {
+    try {
```

Comment:
not sure if this is an issue yet: do we need to go through `throw-try-catch` rather than `switch-case`? `throw` will lock the global JVM lock to get trace stack info, which is considered expensive.